omegaconf/basecontainer.py

```diff
@@ -298,9 +298,15 @@ def expand(node: Container) -> None:
                 else:
                     node._set_value(type_)
 
-        if src._is_missing() and not dest._is_missing():
-            if is_structured_config(get_ref_type(src)):
-                expand(src)
+        if (
+            src._is_missing()
+            and not dest._is_missing()
+            and is_structured_config(get_ref_type(src))
+        ):
+            expand(src)
+            # Ensure all fields of `src` are missing to avoid overwriting `dest`.
+            for k in src:
+                src[k] = MISSING
```

Comment:
Not the cleanest solution but not terrible either.
It would prefer calling get_structured_config_data() to get the dict, fix it, and then use that to create the DictConfig.
See if you can make it work, if it's too hard we can go with this version.

Reply:
I'm not entirely sure I understood what you meant, but I gave it a shot in 7bed03b
Let me know what you think. Personally I find it harder to understand, and I have one concern with my `v._set_value(MISSING)` which I hope isn't overwriting the value of a node that someone might use later (?)

Reply:
Yes, this more efficient because it doesn't need to build the DictConfig and then delete fields.

> v._set_value(MISSING) which I hope isn't overwriting the value of a node that someone might use later (?)

I think in general we are interested in the types and not the values if the containing field is MISSING.
You can try to not expand at all and see which cases are failing. Using expand there was just convenient, but maybe there is a cleaner way to get the same behavior.

Reply:
After taking a second look, not exactly what I had in mind. I am not sure my suggestion would work.

Reply:
> Yes, this more efficient because it doesn't need to build the DictConfig and then delete fields.

I haven't looked closely at `get_structured_config_data()` but since it's returning both keys and values, it seems to me that it's essentially building the same content as a `DictConfig` (except that it's packaged in a regular `dict` rather than a `DictConfig`). Am I missing something?

> > v._set_value(MISSING) which I hope isn't overwriting the value of a node that someone might use later (?)
> 
> I think in general we are interested in the types and not the values if the containing field is MISSING.

Just to clarify my concern: I'm not entirely sure if `v` (a node returned by `get_structured_config_data()`) is a copy that I can safely modify in-place, or may be shared (in other words, would calling again `get_structured_config_data()` return a different `v`, or the same? if it's the same then setting it to `MISSING` is probably unsafe)

> You can try to not expand at all and see which cases are failing. Using expand there was just convenient, but maybe there is a cleaner way to get the same behavior.

This new version isn't actually calling `expand()` anymore (though under the hood it probably does something relatively simliar).

At the end of the day, if `src` is missing but is a structured config type, we still want to make sure that `dest` has all fields for this type. And, for fields that were not already in `dest`, we want them to be missing, but with the correct type. This means creating a node with the proper type and the value set to `MISSING`. Right now we do it in `src`, then merging `src` into `dest`. We could probably do it directly in `dest` but I'm not sure we would actually gain something by it.

I don't really have a better idea right now.

Reply:
> > Yes, this more efficient because it doesn't need to build the DictConfig and then delete fields.
> 
> I haven't looked closely at `get_structured_config_data()` but since it's returning both keys and values, it seems to me that it's essentially building the same content as a `DictConfig` (except that it's packaged in a regular `dict` rather than a `DictConfig`). Am I missing something?

I think you are correct, but in addition to that, you are constructing an instance (which may blow for anything with nested dataclasses), which in turn will have get_structured_config_data() called on it again, and then construct the final DictConfig object.
each of those steps is expensive.

> > > v._set_value(MISSING) which I hope isn't overwriting the value of a node that someone might use later (?)
> > 
> > 
> > I think in general we are interested in the types and not the values if the containing field is MISSING.
> 
> Just to clarify my concern: I'm not entirely sure if `v` (a node returned by `get_structured_config_data()`) is a copy that I can safely modify in-place, or may be shared (in other words, would calling again `get_structured_config_data()` return a different `v`, or the same? if it's the same then setting it to `MISSING` is probably unsafe)

When you call get_structured_config_data(), the result is your to mangle.
 
> > You can try to not expand at all and see which cases are failing. Using expand there was just convenient, but maybe there is a cleaner way to get the same behavior.
> 
> This new version isn't actually calling `expand()` anymore (though under the hood it probably does something relatively simliar).
> 

Expand is doing the same in a different way.

> At the end of the day, if `src` is missing but is a structured config type, we still want to make sure that `dest` has all fields for this type. And, for fields that were not already in `dest`, we want them to be missing, but with the correct type. This means creating a node with the proper type and the value set to `MISSING`. Right now we do it in `src`, then merging `src` into `dest`. We could probably do it directly in `dest` but I'm not sure we would actually gain something by it.
> 
> I don't really have a better idea right now.

IIRC, the reason expand is there is to support cases where you are merging into a MISSING field that is a dataclass.
we want the to expand the field to that any default values in that field are instantiated into the config and the type information in it is created.

I need to take a second look to see why expand is helping here and really get to the bottom of what we are trying to do in the new user case.

Reply:
> > > Yes, this more efficient because it doesn't need to build the DictConfig and then delete fields.
> > 
> > 
> > I haven't looked closely at `get_structured_config_data()` but since it's returning both keys and values, it seems to me that it's essentially building the same content as a `DictConfig` (except that it's packaged in a regular `dict` rather than a `DictConfig`). Am I missing something?
> 
> I think you are correct, but in addition to that, you are constructing an instance (which may blow for anything with nested dataclasses), which in turn will have get_structured_config_data() called on it again, and then construct the final DictConfig object.
> each of those steps is expensive.

But `get_structured_config_data()` is also constructing nested objects as needed -- it seems to me we're only saving the final `DictConfig` creation.

About the rest: ok, thanks!

Reply:
The object it's constructing is not the final one but an intermediate one that is being converted. you are happy to trace the code and see it if yo don't believe me :).

Reply:
> The object it's constructing is not the final one but an intermediate one that is being converted. you are happy to trace the code and see it if yo don't believe me :).

Going to trust you on that :)